### ISSUE TITLE
runtime: enable leak sanitizer in Go

### ISSUE THREAD

**Issue by znkr** [golang/go#67833]
### Proposal Details

This is follow-up to #44853 which proposed to enable ASAN for Go.

One type of error in Go / C interop are memory leaks. These can be detected by ASAN using `ASAN_OPTIONS=detect_leaks=1`. Unfortunately, this doesn't fully work with Go's ASAN integration. For example, allocating a C object and storing it in a global variable in Go is detected as a leak.

I believe to make LSAN work, all we need to do is to tell it about memory regions that Go manages to consider as roots using `__lsan_register_root_region`.

**Comment by ianlancetaylor** [comment]
I don't think this has to be a proposal, taking it out of the proposal process.  We should just do this as part of our ordinary sanitizer support.

It sounds like this will be pretty loose, unless there is some way to tell the sanitizer which pointers are live.  But a conservative leak detector is still better than no leak detector at all.

**Comment by znkr** [comment]
I took another look into LSAN, to understand how conservative it would be. https://maskray.me/blog/2023-02-12-all-about-leak-sanitizer was quite helpful. What that article doesn't document is the `use_poisoned` flag.  IIUC, LSAN can use ASAN poisoning information when scanning for pointers and will ignore pointers in poisoned memory. I think that Go already poisons freed memory, so it might be possible to make this pretty accurate.

**Comment by ruyi789** [comment]
It's a completely redundant feature, the example code doesn't detect the array size at all, and of course it's also your freedom, so why should you be reminded of it

**Comment by qmuntal** [comment]
Enabling the leak sanitizer would be a nice improvement. Note that some ([see list](https://github.com/search?q=__lsan_do_leak_check+language%3AGo&type=code) Go projects are already sidestepping this limitation by manually calling `__lsan_do_leak_check` at program exit.

**Comment by ianlancetaylor** [comment]
Does anybody happen to know which versions of clang and GCC added support for `__lsan_do_leak_check`? Can we just assume it is present?

**Comment by qmuntal** [comment]
> Does anybody happen to know which versions of clang and GCC added support for __lsan_do_leak_check? Can we just assume it is present?

LLVM introduced it in version 3.4 ([source](https://github.com/llvm/llvm-project/commit/655bd0d08ca6df5898c9e64bd00ea39841932edb)), which was released in 2015. GCC followed suite in version 5 ([source](https://github.com/gcc-mirror/gcc/blame/95fba530b6f68f336090abb5699ae9f24d1e22e6/libsanitizer/include/sanitizer/lsan_interface.h#L50)), released also in 2015.

I've been using `__lsan_do_leak_check` for a while with great results, catching real memory leaks in production. You can see its useful by executing the following sample with `go run -asan .`:

```go
package main

// void __lsan_do_leak_check(void);
import "C"

func main() {
	defer C.__lsan_do_leak_check()
	C.CString("foo")
}
```

Which on my machine outputs:

```cmd

=================================================================
==13248==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 4 byte(s) in 1 object(s) allocated from:
    #0 0x7f127460a24f in malloc (/usr/lib/libasan.so.8+0xe124f) (BuildId: ccf997b37481b6e7b757519a36b110bee0a850cd)
    #1 0x472c7b  (/tmp/go-build1729442831/b001/exe/test+0x472c7b) (BuildId: 2adb13e01facb758d13955e5fc42872f5ec67f1f)

SUMMARY: AddressSanitizer: 4 byte(s) leaked in 1 allocation(s).
exit status 1
```

Note that in theory `__lsan_do_leak_check` is automatically called on exit ([source](https://github.com/llvm/llvm-project/blob/af64f0a6c2e26b3fd1979c1fa380136e5528c9b3/compiler-rt/lib/asan/asan_posix.cpp#L192)) if the program is built with `-fsanitize=leak`, but that doesn't work for Go apps.
